models/src/main/scala/coop/rchain/models/PrettyPrinter.scala

```diff
@@ -9,220 +9,180 @@ import coop.rchain.models.Var.VarInstance
-      case BoundVar(level) =>
-        print(s"x${level}")
-      case Wildcard(_) =>
-        print("_")
+      case FreeVar(level)  => s"x${level + freeShift}"
+      case BoundVar(level) => s"x${boundShift - level - 1}"
+      case Wildcard(_)     => "_"
       // TODO: Figure out if we can prevent ScalaPB from generating
-      case VarInstance.Empty =>
-        print("@Nil")
+      case VarInstance.Empty => "@Nil"
     }
 
-  def prettyPrint(c: Channel): Unit =
+  def buildString(c: Channel): String =
     c.channelInstance match {
       case Quote(p) =>
-        print("@{ ")
-        prettyPrint(p)
-        print(" }")
-      case ChanVar(cv) =>
-        prettyPrint(cv)
+        "@{ " + buildString(p) + " }"
+      case ChanVar(cv) => buildString(cv)
       // TODO: Figure out if we can prevent ScalaPB from generating
-      case ChannelInstance.Empty =>
-        print("@Nil")
+      case ChannelInstance.Empty => "@Nil"
     }
 
-  def prettyPrint(t: GeneratedMessage): Unit =
+  def buildString(t: GeneratedMessage): String =
     t match {
-      case v: Var     => prettyPrint(v)
-      case c: Channel => prettyPrint(c)
+      case v: Var     => buildString(v)
+      case c: Channel => buildString(c)
       case s: Send =>
-        prettyPrint(s.chan.get)
-        print("!(")
-        for { (datum, i) <- s.data.zipWithIndex } yield {
-          prettyPrint(datum)
-          if (i != s.data.length - 1)
-            print(",")
-        }
-        print(")")
+        buildString(s.chan.get) + {
+          if (s.persistent) "!!("
+          else "!("
+        } + {
+          ("" /: s.data.zipWithIndex) {
+            case (string, (datum, i)) =>
+              string + buildString(datum) + {
+                if (i != s.data.length - 1) ", "
+                else ""
+              }
+          }
+        } + ")"
+
       case r: Receive =>
-        print("for (")
-        for { (bind, i) <- r.binds.zipWithIndex } yield {
-          printSeq(bind.patterns)
-          print(" <- ")
-          prettyPrint(bind.source.get)
-          if (i != r.binds.length - 1) {
-            print("; ")
+        "for( " + {
+          ("" /: r.binds.zipWithIndex) {
+            case (string, (bind, i)) =>
+              // Stepping past a channel increments boundShift by 1.
+              // Entering a pattern sets freeShift = boundShift and boundShift = 0.
+              string + PrettyPrinter(freeShift, boundShift + i).buildPattern(bind.patterns) + {
+                if (r.persistent) " <= " + buildString(bind.source.get)
+                else " <- " + buildString(bind.source.get)
+              } + {
+                if (i != r.binds.length - 1) " ; "
+                else ""
+              }
           }
-        }
-        print(") { ")
-        prettyPrint(r.body.get)
-        print(" }")
-      case e: Eval =>
-        print("*")
-        prettyPrint(e.channel.get)
+        } + " ) { " + PrettyPrinter(freeShift, boundShift + r.binds.length)
```

Comment:
boundShift should change by
Sum over channels:
  Sum over patterns:
    freecount

Reply:
I don't understand this comment - I think because of the formatting.

Reply:
boundShift should change by receive.bindCount, not receive.binds.length.

receive.bindCount is the sum over all the binds of the sum over all the patterns in a bind of the freeCount.